Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -462,100 +462,71 @@ Most of the frameworks have a parameter to configure the session timeout. This p
-
-##### OWASP MASVS
-* 4.8: "A second factor of authentication exists at the remote endpoint and the 2FA requirement is consistently enforced."
-
-##### CWE
-
--- TODO [Add relevant CWE for "Testing 2-Factor Authentication"] --
-- CWE-312 - Cleartext Storage of Sensitive Information
-
-##### Info
-
-- [1] Meyer's Recipe for Tomato Soup - http://www.finecooking.com/recipes/meyers-classic-tomato-soup.aspx
-- [2] Another Informational Article - http://www.securityfans.com/informational_article.html
-
-##### Tools
-
--- TODO [Add relevant tools for "Testing 2-Factor Authentication"] --
-* Enjarify - https://github.com/google/enjarify
 
+#### Static Analysis
 
+When server-side source code is available, first identify how a second factor or step-up authentication is used and enforced. Afterwards locate all endpoints with sensitive and privileged information and functions: they are the ones that need to be protected. Prior to accessing any item, the application must make sure the user has already passed 2FA or the step-up authentication and that he is allowed to access the endpoint.
 
+2FA or step-up authentication shouldn't be implemented from scratch, instead they should be build on top of available libraries that offer this functionality. The libraries used on the client and server side should be identified and the usage of the available APIs/functions should be verified if they are used accordingly to best practices. Widely used OTP mechanisms are for example:
 
-### Testing Step-up Authentication
+- Google Authenticator
+- Microsoft Authenticator
+- Authy
 
-#### Overview
+Look for keywords in the server source code that are used to implement 2FA or step-up authentication.
 
--- TODO [Provide a general description of the issue "Testing Step-up Authentication".] --
+#### Dynamic Analysis
 
-#### Static Analysis
+First, all privileged endpoints a user can only access with step-up authentication or 2FA within an app should be explored. For all of these requests sent to an endpoint, an interception proxy can be used to capture network traffic. Then, try to replay requests with a token or session information that hasn't been elevated yet via 2FA or step-up authentication. If the endpoint is still sending back the requested data, that should only be available after 2FA or step-up authentication, authentication checks are not implemented properly on the endpoint.
 
--- TODO [Describe how to assess this given either the source code or installer package (APK/IPA/etc.), but without running the app. Tailor this to the general situation (e.g., in some situations, having the decompiled classes is just as good as having the original source, in others it might make a bigger difference). If required, include a subsection about how to test with or without the original sources.] --
+The recorded requests should also be replayed without providing any authentication information, in order to check for a complete bypass of authentication mechanisms.
```

Comment:
Maybe you can provide a few more details/refrences on how to attack the second factor itself as well given the most used factors? (SMS, copy-pasting OTPs from other apps, etc).